Fix double fault and failing mmap

marcrittinghaus opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [ ] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [ ] Tested your changes against relevant architectures and platforms;
 - [X] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [`x86_64`]
 - Platform(s): [`kvm`]
 - Application(s): [N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

This PR fixes two problems with `posix-mmap` and `ukvmem` (see commit messages). The PR should close #786 and #787

marcrittinghaus committed:
plat/kvm-x86: Use IST for regular traps

Although using IST does not allow nested exceptions we go back to
a dedicated stack for regular traps and use IST to switch to it. This
prevents an issue with ukvmem, where a stack is allocated from pageable
heap and pushing to a non-present stack page leads to a double fault.

With this change the stack is switched to the dedicated trap stack and
the page fault for the original stack can be resolved.

A solution to allow nested exceptions would be to reconfigure the trap
to not use IST before executing code that can potentially fail and
restoring IST configuration on afterwards.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
lib/posix-mmap: Fix missing address update

The mmap() call does not return the allocated virtual address via the
addr argument.

Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

csvancea commented:
For me, it fixes the double fault when trying to run `app-elfloader`.

However, calling `mmap` with `flags=MAP_PRIVATE` and `fd!=-1` fails with return code `-ENOTSUP`. The root cause is here:
https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/lib/posix-mmap/mmap.c#L125-L126

Note that `MAP_PRIVATE` is defined as `2`, while `MAP_SHARED_VALIDATE` is defined as `3`. If `flags=MAP_PRIVATE`, the bitwise AND operator returns a non-null value and therefore `UK_VMA_FILE_SHARED` is set.

Later on the call will fail here with the aforementioned error code:
https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/lib/ukvmem/vma_file.c#L47-L48

Changing `(flags & MAP_SHARED_VALIDATE)` to `(flags & MAP_SHARED_VALIDATE) == MAP_SHARED_VALIDATE` fixes the issue.

Edit: Or, `if (flags & MAP_SHARED) {...}` should be enough, provided that the `MAP_*` constants remain unchanged

marcrittinghaus committed:
lib/posix-mmap: Fix MAP_SHARED_VALIDATE

MAP_SHARED_VALIDATE is defined as 3, while MAP_PRIVATE is defined
as 2. A MAP_PRIVATE will thus also evaluate to true for the
MAP_SHARED_VALIDATE test. This commit fixes the flag test.

Co-authored-by: Cosmin Vancea <csvancea@gmail.com>
Co-authored-by: Ioan-Teodor Teugea <ioan_teodor.teugea@stud.acs.upb.ro>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus commented:
Hey @csvancea,

thanks for checking it out and also pointing out the other bug! 😃 How could they use `3` as a flag 😆 I added a commit that includes your suggested change. Thanks for that! Please let me know what E-Mail address I should use for the coauthored tag.

csvancea commented:
Hey @marcrittinghaus,

You can use csvancea@gmail.com as an email address.

And regarding the odd choice, I presume it was driven by the fact that the number of possible flags is very limited, and most of them are already in use. Since MAP_SHARED (1) and MAP_PRIVATE (2) are mutually exclusive, the authors were able to "sneak in" an additional flag without having to reserve (or waste) another bit :smile:

Later edit: There is one more place where MAP_SHARED_VALIDATE is used: https://github.com/unikraft/unikraft/blob/bf44f61631e9d315dcf7c5a3b7b2b53321d51c87/lib/posix-mmap/mmap.c#L126-L127

marcrittinghaus commented:
Hey @John-Ted, @StefanJum, @csvancea,

I updated this PR to also include the second MAP_SHARED_VALIDATE. Since you already tested the commits with PR #799, @StefanJum and @csvancea, can you add your review tags here, too? We can then merge this. Afterwards, @John-Ted can remove the merged commits from #799.

John-Ted commented:
For me, it fixes the mmap double fault on `app-elfloader`, but when trying to read or write to an invalid address, it goes into an infinite loop when trying to read/write to an invalid address.
For the following example, this happens at the line `*f = 'a';`, and if I comment that line, then this happens at the `printf` line:
```C
#include <stdio.h>

int main() {
        char *f = 0x5f;
        *f = 'a';
        printf("0x%hhx\n", *f);
        return 0;
}
```

Here is the output:
```
SeaBIOS (version 1.15.0-1)
Booting from ROM...
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
             Epimetheus 0.12.0~bf44f616
[    0.383376] CRIT: [libkvmplat] <trace.c @   41> RIP: 00000004001098a1 CS: 0008
[    0.430898] CRIT: [libkvmplat] <trace.c @   42> RSP: 000000040006fc70 SS: 0010 EFLAGS: 00010202
[    0.487992] CRIT: [libkvmplat] <trace.c @   44> RAX: 000000000000005f RBX: 000000040006fe68 RCX: 0000000400157eab
[    0.550253] CRIT: [libkvmplat] <trace.c @   46> RDX: 000000040006fe68 RSI: 000000040006fe60 RDI: 0000000000000000
[    0.619504] CRIT: [libkvmplat] <trace.c @   48> RBP: 000000040006fc80 R08: 00000004001d3c30 R09: 0000000000000003
[    0.714647] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000060 R11: 0000000000000206 R12: 0000000000000000
[    0.781190] CRIT: [libkvmplat] <trace.c @   52> R13: 000000040006fe60 R14: 00000004001cd530 R15: 0000000000000001
[    0.841474] CRIT: [libkvmplat] <trace.c @   86> base is 0x40006fc80 caller is 0x400109cfa
[    0.886413] CRIT: [libkvmplat] <trace.c @   86> base is 0x1 RIP: 0000000000104071 CS: 0008
[    0.935249] CRIT: [libkvmplat] <trace.c @   42> RSP: 00000000001afee0 SS: 0010 EFLAGS: 00010002
[    0.984802] CRIT: [libkvmplat] <trace.c @   44> RAX: 00000000001afdf8 RBX: 0000000000000001 RCX: 0000000000000017
[    1.045451] CRIT: [libkvmplat] <trace.c @   46> RDX: 0000003000000030 RSI: 0000000000000000 RDI: 00000000001705d3
[    1.107497] CRIT: [libkvmplat] <trace.c @   48> RBP: 00000000001afef0 R08: 0000000000000046 R09: 00000000000000ff
[    1.170454] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000002 R11: 0000000000000000 R12: 0000000000000000
[    1.230956] CRIT: [libkvmplat] <trace.c @   52> R13: 00000000001aff50 R14: 00000004001cd530 R15: 0000000000000001
[    1.287711] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.329866] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.370668] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.414165] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.457702] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.499958] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.542653] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.589157] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.633837] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.676281] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.717555] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.766127] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.812224] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.864814] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    1.910796] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    1.960380] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.006233] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.033432] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.053144] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.073372] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.095398] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.127195] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.176729] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.220088] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.267559] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.312225] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.356597] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.399269] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53
[    2.445081] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
[    2.493056] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller iConsole terminated, terminating guest (PID: 177904)...
```

For reference, the code currently in the unikraft repo crashes immediately, without any loop:
```
SeaBIOS (version 1.15.0-1)
Booting from ROM...
[    0.127479] CRIT: [libkvmplat] <traps.c @   84> Unhandled Trap 8 (double fault), error code=0x0
[    0.182522] CRIT: [libkvmplat] <trace.c @   41> RIP: 0000000000165d51 CS: 0008
[    0.219901] CRIT: [libkvmplat] <trace.c @   42> RSP: 0000000400070000 SS: 0010 EFLAGS: 00010206
[    0.270205] CRIT: [libkvmplat] <trace.c @   44> RAX: 0000000000000002 RBX: 0000000000165d54 RCX: 00000000c0000100
[    0.325948] CRIT: [libkvmplat] <trace.c @   46> RDX: 0000000000000005 RSI: 00000005fec8c0b0 RDI: 00000005fec88018
[    0.389369] CRIT: [libkvmplat] <trace.c @   48> RBP: 00000000002aedc0 R08: 000000000016d7ca R09: 0000000000170287
[    0.452773] CRIT: [libkvmplat] <trace.c @   50> R10: 00000005fec8b050 R11: 0000000000000001 R12: 00000005fec87018
[    0.518010] CRIT: [libkvmplat] <trace.c @   52> R13: 00000005fec88018 R14: 0000000000000001 R15: 0000000000000008
[    0.579651] CRIT: [libkvmplat] <traps.c @   90> Crashing
Console terminated, terminating guest (PID: 177771)...
```

This only appears to happen in binary compat mode, a native kernel does not go in an infinite loop:
```
SeaBIOS (version 1.15.0-1)
Booting from ROM...
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
             Epimetheus 0.12.0~1cff83c0
[    0.372587] CRIT: [libkvmplat] <trace.c @   41> RIP: 0000000000140dd1 CS: 0008
[    0.427316] CRIT: [libkvmplat] <trace.c @   42> RSP: 00000000001a6f60 SS: 0010 EFLAGS: 00010246
[    0.475883] CRIT: [libkvmplat] <trace.c @   44> RAX: 0000000000000000 RBX: 0000000000000002 RCX: 0000000000000007
[    0.534863] CRIT: [libkvmplat] <trace.c @   46> RDX: 00000000000003d4 RSI: 0000000000000061 RDI: 0000000000142aca
[    0.591660] CRIT: [libkvmplat] <trace.c @   48> RBP: 00000000001a6fc0 R08: 0000000000000202 R09: 00000000000000ff
[    0.652405] CRIT: [libkvmplat] <trace.c @   50> R10: 00000000000003d8 R11: 0000000000000000 R12: 0000000000190d80
[    0.710465] CRIT: [libkvmplat] <trace.c @   52> R13: 0000000000148098 R14: 000000047f9fc020 R15: 0000000400000000
[    0.767914] CRIT: [libkvmplat] <trace.c @   86> base is 0x1a6fc0 caller is 0x122eaf
[    0.813744] CRIT: [libkvmplat] <trace.c @   86> base is 0x1a6fe0 caller is 0x10f4fd
[    0.861570] CRIT: [libkvmplat] <trace.c @   86> base is 0x1a6ff0 caller is 0
[    0.902464] CRIT: [libkvmplat] <trace.c @   66> 
[    0.929712] CRIT: [libkvmplat] <trace.c @   66> 1a6f50: 60 6f 1a 00 00 00 00 00 80 0d 19 00 00 00 00 00
[    0.986974] CRIT: [libkvmplat] <trace.c @   66> 1a6f60: c0 6f 1a 00 00 00 00 00 cc 2d 12 00 00 00 00 00
[    1.040619] CRIT: [libkvmplat] <trace.c @   66> 1a6f70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
[    1.082098] CRIT: [libkvmplat] <trace.c @   66> 1a6f80: 00 00 00 00 00 00 00 00 00 00 00 00 04 00 00 00
[    1.107076] CRIT: [libkvmplat] <trace.c @   66> 
[    1.121307] CRIT: [libkvmplat] <trace.c @   66> 1a6fb0: ff ff ff ff 00 00 00 00 00 70 1a 00 00 00 00 00
[    1.155290] CRIT: [libkvmplat] <trace.c @   66> 1a6fc0: e0 6f 1a 00 00 00 00 00 af 2e 12 00 00 00 00 00
[    1.200408] CRIT: [libkvmplat] <trace.c @   66> 1a6fd0: 00 00 00 00 00 00 00 00 0c 00 00 00 00 00 00 00
[    1.249515] CRIT: [libkvmplat] <trace.c @   66> 1a6fe0: f0 6f 1a 00 00 00 00 00 fd f4 10 00 00 00 00 00
[    1.297387] CRIT: [libkvmplat] <trace.c @   66> 
[    1.325179] CRIT: [libkvmplat] <trace.c @   66> 140dc0: f3 0f 1e fa 55 be 61 00 00 00 bf ca 2a 14 00 31
[    1.380899] CRIT: [libkvmplat] <trace.c @   66> 140dd0: c0 c6 04 25 5f 00 00 00 61 48 89 e5 e8 df 66 fd
[    1.432460] CRIT: [libkvmplat] <trace.c @   66> 140de0: ff 31 c0 5d c3 66 2e 0f 1f 84 00 00 00 00 00 90
[    1.486350] CRIT: [libkvmplat] <trace.c @   66> 140df0: f3 0f 1e fa 55 bf b8 5a 14 00 31 c0 48 89 e5 e8
[    1.544348] CRIT: [libkvmplat] <traps.c @  113> Crashing
Console terminated, terminating guest (PID: 181009)...

```

StefanJum commented:
> Hey @John-Ted, @StefanJum, @csvancea,
> 
> I updated this PR to also include the second MAP_SHARED_VALIDATE. Since you already tested the commits with PR #799, @StefanJum and @csvancea, can you add your review tags here, too? We can then merge this. Afterwards, @John-Ted can remove the merged commits from #799.

Hi @marcrittinghaus, I was waiting for @John-Ted to post this new issue for you to look into it before I add my tag. I can't seem to reproduce it though, running the same app with the `elf-loader` just crashes with a segfault as expected.

marcrittinghaus commented:
> [    0.383376] CRIT: [libkvmplat] <trace.c @   41> RIP: 00000004001098a1 CS: 0008
> [    0.430898] CRIT: [libkvmplat] <trace.c @   42> RSP: 000000040006fc70 SS: 0010 EFLAGS: 00010202
> [    0.487992] CRIT: [libkvmplat] <trace.c @   44> RAX: 000000000000005f RBX: 000000040006fe68 RCX: 0000000400157eab
> [    0.550253] CRIT: [libkvmplat] <trace.c @   46> RDX: 000000040006fe68 RSI: 000000040006fe60 RDI: 0000000000000000
> [    0.619504] CRIT: [libkvmplat] <trace.c @   48> RBP: 000000040006fc80 R08: 00000004001d3c30 R09: 0000000000000003
> [    0.714647] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000060 R11: 0000000000000206 R12: 0000000000000000
> [    0.781190] CRIT: [libkvmplat] <trace.c @   52> R13: 000000040006fe60 R14: 00000004001cd530 R15: 0000000000000001
> [    0.841474] CRIT: [libkvmplat] <trace.c @   86> base is 0x40006fc80 caller is 0x400109cfa
> [    0.886413] CRIT: [libkvmplat] <trace.c @   86> base is 0x1 RIP: 0000000000104071 CS: 0008
> [    0.935249] CRIT: [libkvmplat] <trace.c @   42> RSP: 00000000001afee0 SS: 0010 EFLAGS: 00010002
> [    0.984802] CRIT: [libkvmplat] <trace.c @   44> RAX: 00000000001afdf8 RBX: 0000000000000001 RCX: 0000000000000017
> [    1.045451] CRIT: [libkvmplat] <trace.c @   46> RDX: 0000003000000030 RSI: 0000000000000000 RDI: 00000000001705d3
> [    1.107497] CRIT: [libkvmplat] <trace.c @   48> RBP: 00000000001afef0 R08: 0000000000000046 R09: 00000000000000ff
> [    1.170454] CRIT: [libkvmplat] <trace.c @   50> R10: 0000000000000002 R11: 0000000000000000 R12: 0000000000000000
> [    1.230956] CRIT: [libkvmplat] <trace.c @   52> R13: 00000000001aff50 R14: 00000004001cd530 R15: 0000000000000001
> [    1.287711] CRIT: [libkvmplat] <trace.c @   86> base is 0x1afef0 caller is 0x104f22
> [    1.329866] CRIT: [libkvmplat] <trace.c @   86> base is 0x1aff40 caller is 0x10ee53

It looks like there is a nested page fault. As explained in the commit message of the IST commit this will corrupt the trap stack. With IST the CPU is always setting the stack pointer to the beginning of the trap stack, when a page fault happens. So if you are already in the page fault trap and then access an invalid address the CPU resets the stack pointer. Since the trap stack is active during the second fault, the stack unwinding will try to unwind the (corrupted) trap stack itself.

A solution would be to disable IST whenever we potentially have a nested exception (e.g., when calling `uknofault` from a trap handler). But I would handle that as part of a different PR that introduces a general nested exception support.

StefanJum reviewed (approved):
Got it, then it's all good on my side.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

John-Ted reviewed (approved):
I didn't find any issues other than what was discussed. Looks good to me.
Reviewed-by: Ioan-Teodor Teugea teodor.teugea@gmail.com

csvancea reviewed (approved):
Thanks, @marcrittinghaus!

Reviewed-by: Cosmin Vancea csvancea@gmail.com

marcrittinghaus commented:
Just fixed a typo (`thread` -> `stack`) in the comment.

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
